sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -249,6 +249,12 @@ accessors. USM pointers also convey data dependencies, however offsets into
 system allocations (`malloc`/`new`) are not supported.
 |===
 
+==== Sub-Graph
+
+A node in a graph can take the form of a nested sub-graph. This occurs when
+a command-group submission that invokes `handler::exec_graph()` with an
+executable graph object is added to the graph as a node.
```

Comment:
This would allows us to nest an executable graph within a modifiable graph. Why restricting it to `executable` only?  It might be a cleaner workflow of nesting a modifiable graph within a modifiable graph. That way we could allow optimizations across parent and child graph, and avoid mismatching context while closing both objects separately. To enable ` modifiable` sub-graphs we could overload the `add` function.

Reply:
I don't think only allowing a child graph to be an executable graph has to restrict optimizations. The PIMPL pattern means that an executable & modifiable graph could be implemented the same under the hood. In which case, an executable graph object wouldn't need to have any less information about a graph than the modifiable graph object it was created from. So when a parent graph is being finalized, it could look at that graph information in the child graph to optimize with, regardless of whether that child graph is modifiable or executable. This is how we implemented executable child-graphs in ComputeCpp.

My main concern with modifiable child graphs is if the user then later adds nodes to that child graph, we'd need to define the implications for the parent graph. To get around that, I think we'd have to clone the child modifiable graph as part of the overloaded `add` function. I think that's acceptable though, and is what CUDA graphs does when adding a child graph node.

It doesn't have to be one or the other, we could have both. Overload the `add` function for modifiable child graphs, and
still have `handler::exec_graph()` which can be used to capture executable child graphs without having to do a copy.

Reply:
Yes agreed, both approaches can coexists. I'm fine with merging your patch first. We should carefully analyze the resulting workflow for both modes explicit as well as record&replay, to see if there are any gaps. And if we need the `sub_graph` creation in both states.

> My main concern with modifiable child graphs is if the user then later adds nodes to that child graph, we'd need 
> to define the implications for the parent graph. To get around that, I think we'd have to clone the child modifiable 
> graph as part of the overloaded add function. I think that's acceptable though, and is what CUDA graphs does 
> when adding a child graph node.

Yes I think you're right, we would have to clone the child graph in that case.

Reply:
Cool, i've captured in issue https://github.com/reble/llvm/issues/23#issuecomment-1332034076 that we should research if not having a way to set a modifiable sub-graph is a gap.